### PR TITLE
Added expand network feature

### PR DESCRIPTION
**Pull request by kogeler**
* Deleted iroha_init_vars, iroha_generate_init_configs, iroha_update_runtime_configs configurations variables. The role has become idempotent
* Added custom configuration for docker (environment variables, labels)
* Added debug log level
* If a new node is missing, it will be created and added to the network

### CODE REVIEW

File: ansible/roles/iroha-docker/defaults/main.yml [hunk 1]
```diff
@@ -100,3 +92,16 @@ iroha_docker_logging_max_file: 3
 ## Iroha RAM limit
 # Maximum amount of RAM, it only limits the docker container (e.g. '1024m')
 iroha_docker_mem_limit: ''
+
+##  Iroha logs
+# Enable debug log
+iroha_log_debug: Flase
+
+## Iroha service
+# Iroha commands run on host which has 'iroha_service_host' variable set to true
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/defaults/main.yml]:
```suggestion
# Iroha commands run on host which has 'iroha_service_host' variable set to true
# You can set this variable to any node that you have 
```

File: ansible/roles/iroha-docker/templates/config.docker.j2 [hunk 2]
```diff
@@ -1,4 +1,12 @@
 {
+{% if iroha_log_debug %}
+  "log": {
+      "level": "debug",
+      "patterns": {
+        "trace": "[%Y-%m-%d %H:%M:%S.%F][th:%t][%l] %n %v"
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/templates/config.docker.j2]: we need to check that this pattern will be supported by our elk https://soramitsu.atlassian.net/browse/DOPS-557

File: ansible/roles/iroha-docker/defaults/main.yml (hunk 1 above)
Comment (BulatSaif) [reply ansible/roles/iroha-docker/defaults/main.yml]: Also can you add comment that this setting is not required if you do not change the number of nodes ?

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 3]
```diff
@@ -1,3 +1,89 @@
...
+  - name: New nodes
+    debug:
+      var: iroha_all_new_nodes
+    run_once: yes
+
+  - name: set iroha_old_nodes variable
+    set_fact:
+      iroha_old_nodes: "{{ iroha_nodes | difference(iroha_new_nodes) }}"
+
+  - name: set node_change variable
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: set node_change variable 1
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 4]
```diff
@@ -1,3 +1,89 @@
...
+    set_fact:
+      node_change: true
+    when: iroha_all_new_nodes and (( iroha_all_new_nodes | length ) < ( iroha_all_nodes | length ))
+    run_once: yes
+
+  - name: set node_change variable
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: set node_change variable 2
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml (hunk 3 above)
Comment (BulatSaif) [reply ansible/roles/iroha-docker/tasks/config-gen.yml]: do not use the same task name

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 5]
```diff
@@ -1,3 +1,89 @@
...
+      var: iroha_all_new_nodes
+    run_once: yes
+
+  - name: set iroha_old_nodes variable
+    set_fact:
+      iroha_old_nodes: "{{ iroha_nodes | difference(iroha_new_nodes) }}"
+
+  - name: set node_change variable
+    set_fact:
+      node_change: true
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]: or you can leave only one task, I did not test it, but is should work 
```suggestion
      node_change: "{{ iroha_all_new_nodes and (( iroha_all_new_nodes | length ) < ( iroha_all_nodes | length )) }}"
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 6]
```diff
@@ -17,25 +103,46 @@
     file:
       path: "{{ iroha_config_local_dir }}/{{ inventory_hostname }}/conf/{{ item.human_hostname }}"
       state: directory
-    loop: "{{ iroha_nodes }}"
+    loop: "{{ iroha_new_nodes }}"
+
+  - name: add pub keys to iroha_new_nodes variable
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: add pub keys to iroha_new_nodes variable 1
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 7]
```diff
@@ -17,25 +103,46 @@
...
+    set_fact:
+      iroha_all_new_nodes_temp: "{{ (iroha_all_new_nodes_temp | default([])) + [item | combine({'pub_key': lookup('file', iroha_config_local_dir + '/' + item.hostname + '.pub')})] }}"
+    loop: "{{ iroha_all_new_nodes }}"
+    when: iroha_all_new_nodes and node_change
+    run_once: yes
+
+  - name: add pub keys to iroha_new_nodes variable
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: add pub keys to iroha_new_nodes variable 2
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 8]
```diff
@@ -17,25 +103,46 @@
...
+    run_once: yes
 
   - name: move keys
     copy:
       src: "{{ iroha_config_local_dir }}/{{ item[0].hostname }}{{ item[1] }}"
       dest: "{{ iroha_config_local_dir }}/{{ inventory_hostname }}/conf/{{ item[0].human_hostname }}"
-    loop: "{{ iroha_nodes | product(['.priv', '.pub']) | list }}"
+    loop: "{{ iroha_new_nodes | product(['.priv', '.pub']) | list }}"
 
   - name: move genesis.block
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: move genesis.block 1
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 9]
```diff
@@ -17,25 +103,46 @@
...
     copy:
       src: "{{ iroha_config_local_dir }}/genesis.block"
       dest: "{{ iroha_config_local_dir }}/{{ inventory_hostname }}/conf/{{ item.human_hostname }}"
-    loop: "{{ iroha_nodes }}"
+    loop: "{{ iroha_new_nodes }}"
+    when: not node_change
+
+  - name: move genesis.block
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: move genesis.block 2
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 10]
```diff
@@ -17,25 +103,46 @@
...
+    copy:
+      content: "{{ (iroha_command_register.ansible_facts.docker_container.Output | from_json).result | to_nice_json }}"
+      dest: "{{ iroha_config_local_dir }}/{{ inventory_hostname }}/conf/{{ item.human_hostname }}/genesis.block"
+    loop: "{{ iroha_new_nodes }}"
+    when: node_change
 
   - name: cleanup
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: cleanup 1
```

File: ansible/roles/iroha-docker/tasks/config-gen.yml [hunk 11]
```diff
@@ -17,25 +103,46 @@
...
     file:
       path: "{{ iroha_config_local_dir }}/{{ item[0].hostname }}{{ item[1] }}"
       state: absent
-    loop: "{{ iroha_nodes | product(['.priv', '.pub']) | list }}"
+    loop: "{{ iroha_new_nodes | product(['.priv', '.pub']) | list }}"
 
   - name: cleanup
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/tasks/config-gen.yml]:
```suggestion
  - name: cleanup 2
```

File: ansible/roles/iroha-docker/templates/docker-compose.yml.j2 [hunk 12]
```diff
@@ -3,22 +3,32 @@ version: "2.4"
 services:
 {% for node in iroha_nodes %}
   {{ node.human_hostname }}:
-    image: hyperledger/iroha:{{ iroha_docker_tag }}
+#    image: hyperledger/iroha:{{ iroha_docker_tag }}
+    image: soramitsu/iroha:1.0.1-vote-log
```
Comment (BulatSaif) [inline ansible/roles/iroha-docker/templates/docker-compose.yml.j2]: do not forget to remove

**Comment by BulatSaif** [comment]
Will we implement remove pear logic?

**Comment by igor-egorov** [comment]
:eyes:

**Comment by bakhtin** [comment]
Wait for AddPeer command is fixed in Iroha.

**Pull request closed without merging**